aspnetcore/security/data-protection/consumer-apis/password-hashing/samples/6.x/passwordhasher.cs

```diff
@@ -1,29 +1,24 @@
-using System.Security.Cryptography;
 using Microsoft.AspNetCore.Cryptography.KeyDerivation;
-#nullable disable
+using System.Security.Cryptography;
+
+Console.Write("Enter a password: ");
+string? password = Console.ReadLine();
 
-public class Program
-{
-    public static void Main(string[] args)
-    {
-        Console.Write("Enter a password: ");
-        string password = Console.ReadLine();
+// Generate a 128-bit salt using a cryptographically strong
+// random sequence of nonzero values.
+byte[] salt = new byte[128 / 8];
+RandomNumberGenerator.Create().GetNonZeroBytes(salt);
+Console.WriteLine($"Salt: {Convert.ToBase64String(salt)}");
 
-        // generate a 128-bit salt using a cryptographically strong random sequence of nonzero values
-        byte[] salt = new byte[128 / 8];
-        RandomNumberGenerator.Create().GetNonZeroBytes(salt);
-        Console.WriteLine($"Salt: {Convert.ToBase64String(salt)}");
+// derive a 256-bit subkey (use HMACSHA256 with 100,000 iterations)
+string hashed = Convert.ToBase64String(KeyDerivation.Pbkdf2(
+    password: password!,
+    salt: salt,
+    prf: KeyDerivationPrf.HMACSHA256,
+    iterationCount: 100000,
+    numBytesRequested: 256 / 8));
 
-        // derive a 256-bit subkey (use HMACSHA256 with 100,000 iterations)
-        string hashed = Convert.ToBase64String(KeyDerivation.Pbkdf2(
-            password: password,
-            salt: salt,
-            prf: KeyDerivationPrf.HMACSHA256,
-            iterationCount: 100000,
-            numBytesRequested: 256 / 8));
-        Console.WriteLine($"Hashed: {hashed}");
-    }
-}
+Console.WriteLine($"Hashed: {hashed}");
```

Comment:
This is all that needs a review.

Reply:
@GrabYourPitchforks How do you feel about the non-zero bytes here? I guess there's the possibility, however narrow, that not using it could end up with an entire array filled with zeros, but also, do we care about that tiny reduction in entropy?

Reply:
This shouldn't call GetNonZeroBytes. In fact, we're considering introducing an analyzer that flags these calls as a warning: https://github.com/dotnet/runtime/issues/42763.

More generally, I would need to look into what this sample is trying to accomplish. If this is trying to use PBKDF2 as a mechanism for generating a shared secret key, it's fine, but this wouldn't be an appropriate way to - say - hash a password for storage in a database.